app/Http/Controllers/EventOrdersController.php

```diff
@@ -220,112 +215,131 @@ public function postCancelOrder(Request $request, $order_id)
-                        'amount'               => $refund_amount,
-                        'refundApplicationFee' => floatval($order->booking_fee) > 0 ? true : false,
-                    ]);
-
-                    $response = $request->send();
-
-                    if ($response->isSuccessful()) {
-                        /* Update the event sales volume*/
-                        $order->event->decrement('sales_volume', $refund_amount);
-                        $order->amount_refunded = round(($order->amount_refunded + $refund_amount), 2);
-
-                        if (($order->organiser_amount - $order->amount_refunded) == 0) {
-                            $order->is_refunded = 1;
-                            $order->order_status_id = config('attendize.order_refunded');
-                        } else {
-                            $order->is_partially_refunded = 1;
-                            $order->order_status_id = config('attendize.order_partially_refunded');
-                        }
-                    } else {
-                        $error_message = $response->getMessage();
-                    }
-
-                    $order->save();
-                } catch (\Exeption $e) {
-                    Log::error($e);
-                    $error_message = trans("Controllers.refund_exception");
-                }
+                $errorMessage = trans('Controllers.order_already_refunded');
+            } elseif ($maximumRefundableAmount->isZero()) {
+                $errorMessage = trans('Controllers.nothing_to_refund');
+            } elseif ($refundAmount->isGreaterThan($maximumRefundableAmount)) {
+                // Error if the partial refund tries to refund more than allowed
+                $errorMessage = trans('Controllers.maximum_refund_amount', [
+                    'money' => $maximumRefundableAmount->display(),
+                ]);
             }
+        }
+
+        // Stop before refunding anything or updating stats
+        if ($errorMessage) {
+            return response()->json([
+                'status'  => 'error',
+                'message' => $errorMessage,
+            ]);
+        }
 
-            if ($error_message) {
+        // Do refund with the amount calculated
+        try {
+            $gateway = Omnipay::create($order->payment_gateway->name);
+            $gateway->initialize($order->account->getGateway($order->payment_gateway->id)->config);
+
+            $request = $gateway->refund([
+                'transactionReference' => $order->transaction_id,
+                'amount' => $refundAmount->toFloat(),
+                'refundApplicationFee' => floatval($order->booking_fee) > 0 ? true : false,
+            ]);
+
+            Log::debug("STRIPE REQUEST", [
+                'transactionReference' => $order->transaction_id,
+                'amount' => $refundAmount->toFloat(),
+                'refundApplicationFee' => floatval($order->booking_fee) > 0 ? true : false,
+            ]);
```

Comment:
Hard coding STRIPE REQUEST here is not good as Attendize uses Omnipay so probably better to log the gateway_name